recommender/engines/autoencoders/training/model_training_step/model_training_step.py

```diff
@@ -0,0 +1,338 @@
+                    epoch,
+                )
+                writer.flush()
+
+            tepoch.set_postfix(
+                loss=loss, val_loss=val_loss, best_model=str(best_model_flag)
+            )
+
+    end = time.time()
+    timer = end - start
+
+    if verbose:
+        print(f"Total training time: {end - start}")
+
+    return best_model, timer
+
+
+def perform_training(
+    collection_name,
+    train_ds_dl,
+    embedding_dim,
+    features_dim,
+    writer,
+    device,
+    learning_rate,
+    epochs,
+    verbose,
+):
+    """Perform training"""
+
+    autoencoder_model = create_autoencoder_model(
+        collection_name,
+        features_dim=features_dim,
+        embedding_dim=embedding_dim,
+        writer=writer,
+        train_ds_dl=train_ds_dl,
+        device=device,
+    )
+
+    optimizer = Adam(autoencoder_model.parameters(), lr=learning_rate)
+
+    trained_autoencoder_model, timer = train_autoencoder(
+        model=autoencoder_model,
+        optimizer=optimizer,
+        loss_function=autoencoder_loss_function,
+        epochs=epochs,
+        train_ds_dl=train_ds_dl,
+        writer=writer,
+        save_period=10,
+        verbose=verbose,
+        device=device,
+    )
+
+    loss = evaluate_autoencoder(
+        trained_autoencoder_model,
+        train_ds_dl,
+        autoencoder_loss_function,
+        device,
+    )
+
+    print(f"User Autoencoder testing loss: {loss}")
+
+    return trained_autoencoder_model, loss, timer
+
+
+class AEModelTrainingStep(ModelTrainingStep):
+    """Autoencoder model training step"""
+
+    def __init__(self, pipeline_config):
+        super().__init__(pipeline_config)
+        self.device = self.resolve_constant(DEVICE, torch.device("cpu"))
+        self.writer = self.resolve_constant(WRITER)
+        self.verbose = self.resolve_constant(VERBOSE, True)
+        self.user_batch_size = self.resolve_constant(USER_BATCH_SIZE, 128)
+        self.service_batch_size = self.resolve_constant(SERVICE_BATCH_SIZE, 128)
+        self.user_embedding_dim = self.resolve_constant(USER_EMBEDDING_DIM, 32)
+        self.service_embedding_dim = self.resolve_constant(SERVICE_EMBEDDING_DIM, 64)
+        self.learning_rate = self.resolve_constant(LR, 0.01)
+        self.epochs = self.resolve_constant(EPOCHS, 500)
+
+        self.trained_user_autoencoder_model = None
+        self.trained_services_autoencoder_model = None
+
+    def __call__(self, data=None) -> Tuple[object, dict]:  # TODO Refactor
+        """Perform user and service training"""
+
+        raw_data = data[AUTOENCODERS]
+
+        # Users
+        user_train_ds = raw_data[USERS][TRAIN]
+
+        user_train_ds_dl = DataLoader(
+            user_train_ds, batch_size=self.user_batch_size, shuffle=True
+        )
+
+        user_features_dim = len(
+            User.objects[0].one_hot_tensor
+        )  # TODO change the location of those
+        service_features_dim = len(Service.objects[0].one_hot_tensor)
```

Comment:
This data is produced in the DataPreparationStep. If you need them here you should get them from the data parameter and they should be placed there by the previous step(s). Model Training step shouldn't have any interaction with the DB (There is one exception to this rule: saving a trained model but this operation should take place inside save function as defined in the BaseModelTrainingStep abstract class).